<a href="https://colab.research.google.com/github/omazevych/python_for_da_tasks/blob/main/10_4_apply%2C_groupby%2C_pivot_table.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [2]:
df = pd.read_csv('drive/MyDrive/Data Analysis/Python/homeworks/supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
branch_profit = df.groupby('Branch')['gross income'].sum()
most_profitable_branch = branch_profit.idxmax()
max_profit = branch_profit.max()

print(f"\nНайприбутковіший філіал: {most_profitable_branch} з сумарним прибутком {max_profit:.2f}")


Найприбутковіший філіал: C з сумарним прибутком 5265.18


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [4]:
city_of_branch = df[df['Branch'] == most_profitable_branch]['City'].iloc[0]
print(f"Філіал з найвищим прибутком: {most_profitable_branch}")
print(f"Місто цього філіалу: {city_of_branch}")

Філіал з найвищим прибутком: C
Місто цього філіалу: Naypyitaw


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [9]:
invoices_by_category = df.pivot_table(
    index='Branch',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count'
)
print(invoices_by_category)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45  


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [10]:
branch_max_electronic = invoices_by_category['Electronic accessories'].idxmax()
max_invoices = invoices_by_category['Electronic accessories'].max()

print(f"Філіал з найбільшою кількістю інвойсів у категорії 'Electronic accessories': {branch_max_electronic}")
print(f"Кількість інвойсів у цій категорії: {max_invoices}")

Філіал з найбільшою кількістю інвойсів у категорії 'Electronic accessories': A
Кількість інвойсів у цій категорії: 60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


*Гіпотеза №1: Чи існує кореляція між кількістю товарів у покупці та сумою чеку  у різних категоріях товарів?*  
Це дозволяє зрозуміти, які категорії створюють більші чеки при збільшенні кількості товарів, і іх можна взяти за основу для акцій.

*Гіпотеза №2: Чи впливає день тижня на кількість покупок різних категорій товарів?*  
Знову ж можна запускати цільові акції у певні дні тижня для конкретних категорій товарів, типу акція вихідного дня або ціна тижня по середам.

*Гіпотеза №3: Чи залежить задволеність клієнта від типу філіалу?*  
Філії з низьким рейтингом варті уваги щодо поліпшення обслуговування та оптимізації процесів.

In [13]:
#1
category_corr = df.groupby('Product line')[['Quantity', 'Total']].corr().iloc[0::2,-1].reset_index()
print(category_corr[['Product line','Total']])

             Product line     Total
0  Electronic accessories  0.692608
1     Fashion accessories  0.747908
2      Food and beverages  0.710387
3       Health and beauty  0.684509
4      Home and lifestyle  0.686509
5       Sports and travel  0.707104


In [14]:
#2
df['Date'] = pd.to_datetime(df['Date'])
df['Weekday'] = df['Date'].dt.day_name()
weekday_category_counts = df.groupby(['Weekday', 'Product line'])['Invoice ID'].count().reset_index(name='count')
print(weekday_category_counts)

      Weekday            Product line  count
0      Friday  Electronic accessories     17
1      Friday     Fashion accessories     30
2      Friday      Food and beverages     19
3      Friday       Health and beauty     26
4      Friday      Home and lifestyle     24
5      Friday       Sports and travel     23
6      Monday  Electronic accessories     23
7      Monday     Fashion accessories     18
8      Monday      Food and beverages     22
9      Monday       Health and beauty     26
10     Monday      Home and lifestyle     20
11     Monday       Sports and travel     16
12   Saturday  Electronic accessories     27
13   Saturday     Fashion accessories     35
14   Saturday      Food and beverages     31
15   Saturday       Health and beauty     19
16   Saturday      Home and lifestyle     24
17   Saturday       Sports and travel     28
18     Sunday  Electronic accessories     20
19     Sunday     Fashion accessories     20
20     Sunday      Food and beverages     23
21     Sun

In [15]:
#3
branch_rating = df.groupby('Branch')['Rating'].mean().reset_index()
print(branch_rating)

  Branch    Rating
0      A  7.027059
1      B  6.818072
2      C  7.072866


***рекомендації***

№1. Кореляція показує, наскільки збільшення кількості товарів веде до збільшення суми чеку і оскільки ми маємо помірно‑високу позитивну кореляцію для усіх 5 категорій, можемо запускати акції типу “купи 2 і 3 отримай у подарунок”, щоб стимулювати більші покупки. Можна спрбувати стимулювати допродаж менш дорогих товарів або ж запропонувати комбо-набори, щоб підвищити середній чек.

№2. Планувати цільові маркетингові акції та промо-пропозиції для кожної категорії відповідно до дня тижня, щоб максимізувати продажі та підвищити середній чек. Наприклад:  
- Понеділок та вівторок – акції на косметику, засоби для догляду та здоров’я.
- Середа – варто зробити знижки на продукти харчування, напої, готові набори для дому.
- Четвер – акцій на електроніку, гаджети, аксесуари.
- П’ятниця та субота – акції на одяг, аксесуари, сезонні колекції.
- Неділя – можна робити акції на товари для дому, декор, побутові товари.

№3.  Усі філіали мають рейтинг Medium (<8), необхідно проаналізувати причини такого рейтингу і усунути їх. Також варто відмітити, що філіал B має найнижчий рейтинг (6.82)- він заслуговує уваги в першу чергу, бо є проблема, що не дозволяє поставити вищий рейтинг: рівень обслговування, асортимент чи наявність товврів.

